Issue와 PR 환경 설정

yoouyeon opened the pull request:
## 📝 관련 이슈

- close #2 

## 💻 작업 내용

> #4 내용이 함께 들어 있습니다.

Issue와 PR과 관련된 Github Action들을 추가했습니다.
하다보니까 양이 많아졌네요... ㅜㅅㅜ
최대한 주석을 많이 달아두려고 했는데 혹시 이해 안되는 부분 있으면 코멘트 남겨 주세요!! 😁

### PR Review Manager

> `.github/workflows/pr-review-manager.yml` 파일입니다

#### 1. PR 작성자를 Assignee로 지정하고 Reviewer를 지정하는 작업 (`jobs > assign`)

[review-assign-action](https://github.com/hkusu/review-assign-action) 액션을 이용했습니다.

- PR이 열리거나, Draft 상태에서 ready to review 상태로 전환되었을 때 트리거되는 액션
- Assignee는 PR을 작성한 사람을 Assignee로 지정합니다.
- Reviewer는 Repository Variables 에 정의해 둔 유저들 중에서 Assignee를 제외한 사람들을 지정합니다.
  - 저흰 두명이라 ㅋㅋ 리뷰어 수를 따로 지정해두진 않았어요
  - 정의해 둔 데이터는 레포지토리 Settings > Secrets and Variables > Actions > Variables > Repository variables 에서 확인할 수 있습니다!

#### 2. 연결된 이슈의 라벨을 PR에 추가하는 작업 (`jobs > sync-labels`)

딱 원하는 기능을 하는 배포된 액션을 못찾기도 했고
쉘 스크립트로 작성하기엔 기능이 복잡해서
`.github/actions/sync-labels` 에 JS로 액션을 구현했습니다.
(TS를 쓰기엔 간단한 액션을 만드는 것 뿐인데 부가적으로 설정할 것이 많다고 느껴져서 JS로 했어요.)

```
.
├── action.yml : workflows/pr-review-manager.yml 에서 호출하는 액션 파일
├── dist
│   └── index.js : src/index.js를 빌드한 코드 (매 실행마다 패키지를 설치하지 않아도 되도록 빌드 결과물을 실행하게 함)
├── package.json
├── src
│   └── index.js : 액션 구현 코드
└── yarn.lock
```

**실행 흐름**

1. Checkout Action
    - Github Action 실행 환경에 .github/actions/sync-labels 파일이 있어야 하기 때문에 [checkout](https://github.com/actions/checkout) 액션을 이용해서 가져옵니다.
    - 액션에 필요한 .github/actions/sync-labels 디렉토리만 내려받도록 했습니다.
    - 최신 상태만 필요해서 최신 커밋만 가져오도록 했습니다.
2. Sync labels from linked issues
    - PR이 올라온 feat/#123-branch-name 형식의 **브랜치 이름**에서 관련된 Issue 번호를 가져옵니다.
    - Issue 번호를 이용해서 Issue에 달린 라벨을 가져옵니다.
    - 기존 PR에 달려 있던 라벨과 결합해서 새로운 라벨 목록을 만들고
    - PR에 라벨을 달아줍니다

### 🚨 머지 커밋 앞에 "merge: " 붙여주기

> 🚨 이거 적용은 좀 논의해봐야 할 것 같아요!!

> `.github/workflows/merge-prefixer.yml` 파일입니다.

**실행 흐름**

1. Checkout the target branch
    - Github Action 환경에서 git cli를 사용하기 위해서 [checkout](https://github.com/actions/checkout) 액션을 이용해서 전체 레포지토리를 가져옵니다.
    - PR의 base가 되는 브랜치에 push된 커밋을 수정해야 하기 때문에 base 브랜치로 체크아웃 했습니다.
    - 기존 커밋 히스토리를 모두 가져와서 덮어씌워야 하기 때문에 fetch-depth: 0을 설정했습니다. (참고: https://github.com/actions/checkout?tab=readme-ov-file#Fetch-all-history-for-all-tags-and-branches)
2. Configure Git
    - 기본적으로  설정된 유저 정보가 없더라구요... github 유저 정보를 등록해줬습니다.
    - 개발자가 아닌 github action을 이용해서 수정된 커밋이라는 것을 표현하기 위해서 bot 계정으로 등록했습니다.
        공식적으로 액션에 이 유저를 쓰라는 설명은 못찾았는데, github에서 유지보수하고 있는 [checkout 액션에서 사용하는 예시](https://github.com/actions/checkout?tab=readme-ov-file#push-a-commit-using-the-built-in-token)가 있는 것을 보면 관습적으로 사용하는 유저인 것 같아요.. 🙄
3. Add Prefix
    - 가장 최신의 커밋 메시지를 가져옵니다. (title과 body 포함)
    - 커밋 메시지의 title (가장 윗줄)을 추출합니다.
    - title이 "merge: " 로 시작하면 추가로 prefix를 붙이지 않습니다.
    - sed 명령어를 이용해서 가장 앞줄에 "merge: " 를 붙여줍니다. (참고: https://hmwoo.tistory.com/37)
    - amend 옵션으로 커밋 메시지를 수정하고 강제 푸시합니다 (이미 원격에 올라간 커밋을 수정하는 것이기 때문에)

## ✅ 테스트 리스트

**저희 레포지토리에서 확인해 본 것**

- [x] Assignee 할당
   <img width="682" alt="Screen_Shot 2025-01-21 20 52 03" src="https://github.com/user-attachments/assets/f3d4e240-a464-4431-be9e-1aba1900e136" />
- [x] 라벨 설정 (이슈 O & 라벨 O)
   <img width="682" alt="Screen_Shot 2025-01-21 20 51 55" src="https://github.com/user-attachments/assets/5e19eb74-4545-4679-a82a-1e38b5157d9c" />
- [x] Reviewer 할당 (Draft -> Ready To Review)
   <img width="682" alt="Screen_Shot 2025-01-22 10 12 40" src="https://github.com/user-attachments/assets/5088a754-193b-49da-97ef-78e8b8078e0e" />

**개인 레포지토리에서 확인해 본 것**
- [x] Reviewer 할당 (Open)
   <img width="682" alt="Screen_Shot 2025-01-21 21 16 03" src="https://github.com/user-attachments/assets/292c835a-1fdd-454f-b16f-42f4c59b2f6d" />
- [x] 라벨 설정 (이슈 O & 라벨 X): 이름 warning은 수정했습니다.. ([f610aa2](https://github.com/dnd-side-project/dnd-12th-7-frontend/pull/9/commits/f610aa20801c3ff65127fc3ee4f6031e3d84e848))
   <img width="682" alt="Screen_Shot 2025-01-21 22 05 14" src="https://github.com/user-attachments/assets/460ddf54-0530-44e0-9997-accf78212484" />
- [x] 라벨 설정 (이슈 X)
   <img width="682" alt="Screen_Shot 2025-01-21 21 16 37" src="https://github.com/user-attachments/assets/ce8a30f4-6608-4f43-9f98-3a3a5179370b" />
- [x] merge commit prefix 붙이기
   <img width="682" alt="Screen_Shot 2025-01-21 22 07 11" src="https://github.com/user-attachments/assets/8f947246-0604-473b-8091-95a39de374a8" />
- [x] merge commit prefix 가 존재하면 넘어가기
   <img width="682" alt="Screen_Shot 2025-01-21 22 11 57" src="https://github.com/user-attachments/assets/ba7fe7ee-1946-4f70-820a-8c0b79d581eb" />

## 👻 리뷰 요구사항

merge commit을 수정할 때에 이미 원격에 올라간 commit을 수정하는 것이기 때문에 force push를 사용하고 있는데요...
**직접 하는게 아니라 github action으로 강제 푸시를 하다 보니까 예상치 못한 문제가 발생할 수도 있겠다는 생각이 듭니다.**
(작업 중에는 커밋 히스토리가 다 날아가버린 경우도 경험했거든요 ㅋㅋㅜ)
**문제가 발생하면 복구하기 어려워질 것 같아서 이거는 지금 액션을 적용하기보단 번거로워도 "merge: " 를 직접 붙여주는 것도 좋을 것 같은데
어떤가요..?? 🥲** 아니면 자동으로 붙여주는 것까진 아니더라도 덜 번거로운 방법이 있을까요... 🥹🥹

yoouyeon committed:
docs: PR 템플릿 추가

yoouyeon committed:
docs: issue 템플릿 추가

- 기능 추가 템플릿
- 버그 제보 템플릿

yoouyeon committed:
docs: Bug 템플릿 내용 추가

- 재현 방법 항목 구체화 → 발생 상황 / 예상 결과
- https://github.com/dnd-side-project/dnd-12th-7-frontend/pull/4#pullrequestreview-2560346303

yoouyeon committed:
merge: Issue와 PR 템플릿 추가 (#4)

yoouyeon committed:
chore: PR Assignee와 Reviewer 할당 액션 추가

https://github.com/hkusu/review-assign-action

yoouyeon committed:
comment: assign 작업 설명 주석 추가

yoouyeon committed:
chore: merge commit type prefix 추가 액션 추가

yoouyeon committed:
chore: PR에 연결된 이슈의 라벨을 복사하는 액션 추가

yoouyeon committed:
fix: token 이름 수정

github-token과 token이 혼용되고 있었음

yoouyeon committed:
fix: commit history가 사라지는 문제 수정

- checkout 의 fetch-depth 기본값이 1임
- 최신 커밋 1개만 가져오므로 push --force-with-lease 시에 기존 히스토리가 사라지는 문제가 있었음
- fetch-depth: 0 으로 설정해서 모든 커밋 히스토리를 가져오도록 수정함

ongheong reviewed (approved):
유연님 정말 Github Action을 잘 다루시는군요...! 😮 PR 열 때 똑같은 assignee, reviewer 매번 누르는게 은근 귀찮았는데 assign 액션과 sync-label 액션으로 더 빠릿하게 작업을 진행할 수 있겠네요!

merge를 자동으로 붙여주는 기능은 처음에는 매우 좋다고 생각했는데, 유의점을 다시 보니 force push를 사용하는 방법이었네요...! 좀 더 안전한 방법은 없는지 저도 조금 찾아봤는데, 저는 주로 PR Merge를 웹 GUI에서 하는 편이라 이런 경우에는 접두사를 직접 붙여주는 것 말고 뾰족한 대안이 없네요...

유연님 말대로 이 부분은 자동화로 하지 않고 "merge: "를 직접 붙이는 방법이 나을 것 같습니다!

yoouyeon committed:
remove: "Add Prefix to Merge Commits" 액션 제거

force push의 위험성으로 인해 액션을 제거하기로 함
https://github.com/dnd-side-project/dnd-12th-7-frontend/pull/9#pullrequestreview-2566288311

yoouyeon commented:
@ongheong 저도 웹에서 하는 편이라... 마땅한 방법이 생각나지 않더라구요 ㅎㅎㅜㅜ
https://github.com/dnd-side-project/dnd-12th-7-frontend/pull/9/commits/cf8705a4f373cc45fa7a88ea83186d0549fa34f6 merge 시에 발생하는 액션은 제거했습니다! 번거롭더라도 잊지 말고 `merge: ` 잘 붙여주도록 합시다..ㅎㅎ

저는 이 작업 끝나고 나면 특별히 할 일이 없어서 🥹
여은님 린트랑 다른 패키지들 설정하시는 동안 저는 이번 작업에 이어서 Zenhub도 이슈를 Review 상태로 옮길 수 있는 방법이 있는지 찾아보도록 할게요!!! 💪
